src/main/java/jenkins/plugins/openstack/compute/internal/CachedFunction.java

```diff
@@ -0,0 +1,186 @@
+
+    /** {@inheritDoc} */
+    @Override
+    public V get(final K key) {
+        final Callable<Result> callOnCacheMiss = new Callable<Result>() {
+            @Override
+            public Result call() {
+                try {
+                    return new Result(calculate(key));
+                } catch (RuntimeException e) {
+                    return new Result(e);
+                }
+            }
+        };
+        final Object cacheKey = key == null ? NULL : key;
+        final Result result;
+        try {
+            result = cache.get(cacheKey, callOnCacheMiss);
+        } catch (ExecutionException e) {
+            // Should not happen as our callOnCacheMiss method does not throw
+            // a declared exceptions and any RuntimeExceptions thrown by the
+            // calculate method will have been caught and stored in the result.
+            throw new RuntimeException("Internal error", e);
+        }
+        // If we got an exception, pass that back...
+        if (result.ex != null) {
+            // NOTE: If you are debugging an exception stacktrace and you are
+            // pointed to the following line, please be aware that this is where
+            // get() was called from this time, but all the deeper stacktrace
+            // elements came from the original exception from when the value was
+            // first calculated (and cached), which might not be from this time
+            // if "this time" was a cache-hit.
+            final List<StackTraceElement> howWeGotHereThisTime = createStacktracePointingHere();
+            final StackTraceElement[] st = combineHowWeGotHereWithTheOriginalStacktrace(howWeGotHereThisTime,
+                    result.exOriginalStacktrace);
+            result.ex.setStackTrace(st);
+            throw result.ex;
+        }
+        // ...otherwise return the result.
+        return result.value;
+    }
+
+    /**
+     * When we are returning the "same exception" that we cached previously, we
+     * need to re-throw the exact same exception (same instance) that was thrown
+     * when we calculated the results (in order to ensure that cache-hit
+     * behavior is the same as cache-miss behavior) BUT the stacktrace from the
+     * cached result will be very misleading if left as-is (it'll show us being
+     * called from whatever code experienced the cache-miss), so we take the
+     * exception's original stacktrace (up to where we got involved) and then
+     * glue our current stacktrace (from where we got involved onwards) on to
+     * the end of that.
+     * 
+     * @param howWeGotHereThisTime
+     *            Our current stacktrace up to where we got involved.
+     * @param originalExceptionStacktrace
+     *            The original exception stacktrace.
+     * @return A stacktrace showing the cache-miss code path plus how we got to
+     *         this point.
+     */
+    private static StackTraceElement[] combineHowWeGotHereWithTheOriginalStacktrace(

```

Comment:
I am wondering if we want to cache the exceptions. Would it not be better to cache only the real data so (recoverable) failures are retried? Or I guess we can try to tell what exception types are expected to be recoverable and not cache those.

---

All in all, I have a bad feeling about manipulating stacktraces this way as when we fail to do it right, it will be whole lot more confusing that if we do nothing. But I do not have any better suggestion though. I lean towards indicating the exception is cached not to mislead administrators that they observed more communication errors or the problems lasted longer than they did (ex. one connection refused exception will be reported over and over for up to 15 seconds but because we have hidden the caching just too well, we will make an impression the problem lasted longer, number of requests ware different and individual failures occurred at different times).

Reply:
I think there are good and bad aspects to either approach.

I'm of the opinion that, in general, caching failure is just as important as caching success.  The purpose of this code is to improve UI responsiveness and reduce load on the Openstack server. That means not asking the Openstack server the same question more often than necessary, so if we get an exception the first time then we're likely to get the same exception if we repeat that same question a dozen times in a row (for most things - we don't cache any result where that is not expected to be the case).
So I believe that, in general, caching exceptions is a _good_ thing.
...although, in an ideal world, we'd cache exceptions for less time than we cache the successful results.

That said, if we cache exceptions then avoiding confusion is _hard_.  As you've seen in the code, most of that caching code here is about exception handling - caching _without_ exception handling is dead simple by comparison.
I'm not generally in favor of code that's needlessly complicated - I prefer code to be as simple as possible _but_ no simpler.  I prefer code to be correct more than I prefer code to be simple :-)
I admit that this stacktrace-manipulation code is complicated and nasty, and it would be nice to avoid any need for this kind of thing...

I do wonder if the potential confusion could be avoided if we include a timestamp of _when_ the exception was cached and include that in the exception?
...but I'm not sure how we could do that, as that would require changing the message text of the exception.

An alternative approach would be to stop throwing the original exception but instead to _always_ wrap the original exception, and put our own message in to those exceptions (either stating that the exception happened just now, or if it happened some time ago and is cached).
This would mean that calling the cache would always be different from calling the un-cached function, but as we control both ends then I guess we could accept that.